test/token/ERC20/utils/SafeERC20.test.js

```diff
@@ -144,6 +151,102 @@ describe('SafeERC20', function () {
       });
     });
   });
+
+  describe('with ERC1363', function () {
+    const value = 100n;
+    const data = '0x12345678';
+
+    beforeEach(async function () {
+      this.token = this.erc1363Mock;
+    });
+
+    shouldOnlyRevertOnErrors();
+
+    describe('transferAndCall', function () {
+      it('cannot transferAndCall to an EOA directly', async function () {
+        await this.token.$_mint(this.owner, 100n);
+
+        await expect(this.token.connect(this.owner).transferAndCall(this.receiver, value, ethers.Typed.bytes(data)))
+          .to.be.revertedWithCustomError(this.token, 'ERC1363InvalidReceiver')
+          .withArgs(this.receiver);
+      });
+
+      it('can transferAndCall to an EOA using helper', async function () {
+        await this.token.$_mint(this.mock, value);
+
+        await expect(this.mock.$transferAndCallRelaxed(this.token, this.receiver, value, data))
+          .to.emit(this.token, 'Transfer')
+          .withArgs(this.mock, this.receiver, value);
+      });
+
+      it('can transferAndCall to an ERC1363Receiver using helper', async function () {
+        await this.token.$_mint(this.mock, value);
+
+        await expect(this.mock.$transferAndCallRelaxed(this.token, this.erc1363Receiver, value, data))
+          .to.emit(this.token, 'Transfer')
+          .withArgs(this.mock, this.erc1363Receiver, value)
+          .to.emit(this.erc1363Receiver, 'Received')
+          .withArgs(this.mock, this.mock, value, data);
+      });
+    });
+
+    describe('transferFromAndCall', function () {
+      it('cannot transferFromAndCall to an EOA directly', async function () {
+        await this.token.$_mint(this.owner, value);
+        await this.token.$_approve(this.owner, this.other, ethers.MaxUint256);
+
+        await expect(
+          this.token
+            .connect(this.other)
+            .transferFromAndCall(this.owner, this.receiver, value, ethers.Typed.bytes(data)),
+        )
+          .to.be.revertedWithCustomError(this.token, 'ERC1363InvalidReceiver')
+          .withArgs(this.receiver);
+      });
+
+      it('can transferFromAndCall to an EOA using helper', async function () {
+        await this.token.$_mint(this.owner, value);
+        await this.token.$_approve(this.owner, this.mock, ethers.MaxUint256);
+
+        await expect(this.mock.$transferFromAndCallRelaxed(this.token, this.owner, this.receiver, value, data))
+          .to.emit(this.token, 'Transfer')
+          .withArgs(this.owner, this.receiver, value);
+      });
```

Comment:
This is the happy path, and it's testing a token that returns true on every call. We need a case for a token that returns false and it should revert, and another when the token doesn't return any value and it should pass.

Reply:
We don't have ERC-1363 mocks that return false or that return no data when called. Both would not match the ERC requirements. I'm not sure we have to specifically write non-standard contracts, that are not known to exist already, to test against them.

The idea of this library was to help interract with standard and common tokens. 

For ERC-20:
- "standard" includes tokens that return false instead of reverting (that is allowed)
- "common" includes tokens without return data (that is not in the ERC, but some important contract do that ...)

For ERC-1363:
- "standard" → the ERC explicitelly says that the token should return true or revert. returning false, or not returning anything, is not ok
- "common" → I'm not sure there is any widelly used ERC-1363 that returns false, or returns no data.

So again, I don't think we should support (and test) interractions with broken contracts.

Reply:
Yes, I agree with your stand. But in my opinion the discussion point is around what "common tokens" mean.

Since the ERC20 standard includes tokens returning `false` instead of reverting, it's possible (and very likely imo) that an ERC1363 will end up behaving as in the ERC20 standard by overriding the `transfer` function to return false, either in the first deployment or an upgrade.

See https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4631#discussion_r1459521680